[depends][Android] Add base dependencies setup for libdovi

quietvoid opened the pull request:
## Description
The changes bring support for the Rust toolchain with the goal of being able to integrate C compatible dependencies through [cargo-c](https://github.com/lu-zero/cargo-c).
This PR also adds the optional [libdovi](https://github.com/quietvoid/dovi_tool/tree/main/dolby_vision) dependency for Android builds.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
The main motivation for this is to be able to convert Dolby Vision on-the-fly, so that Android devices have less issues.
This requires `libdovi` (which is a Rust/C compatible library).

There are also other usecases for Rust such as binary addons, which seem to have been previously proposed GSoC projects.

## How has this been tested?
Tested successful build for all Android target platforms.
Runtime tested on FireTV Stick 4K (2018) and 4K Max (2021), so `armv7`.

## What is the effect on users?
None by default, as the added dependencies are built optionally with a configure flag.
Only the build process is affected, it's still unknown to me how the new dependencies behave on all supported platforms.
As the dependencies are limited to Android only, I don't think it can have a negative effect.

Testing thread for Android: https://forum.kodi.tv/showthread.php?tid=371557

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

quietvoid commented:
Opened this as draft as there are very likely changes needed.

fuzzard commented:
I would say you are going to have to implement something in the core code that uses this before it's looked at. 

Will be interesting to see how much the depends build times increase with rustup

fritsch commented:
Maybe - as the stuff is used from the sidedata (when used from mkv container) at the end - a place inside ffmpeg would be great I think, though having it as a customer converter, like e.g. VideoPlayerPassthroughCodec would be great for the start as well - here it could be used as a transcoder, similar to what we do with our Dolby Transcoding.

fuzzard commented:
You'll want to add a "temp" commit that actually enables these flags for this PR so we can actually have them built as well

fritsch commented:
I like that we have a PR here in the first place. Which could stay open or at a point also could point to a dovi public branch in our repo, so that other contributors can help on this basis, so that a future solution that nicely fits into kodi can be found? Chances are higher, when dependencies are already available I'd say.

fuzzard commented:
Yeah, but we aren't ever going to merge a dependency not actively used, and also this doesn't even test the build of it, as default is off.

quietvoid commented:
I could add a simple commit that does nothing but use the `libdovi` dependency, I guess.

quietvoid commented:
I've made the dependency default for Android.
It'll most likely fail x86 targets as `HOST` won't match the [Rust target platforms](https://doc.rust-lang.org/nightly/rustc/platform-support.html).

quietvoid commented:
This is the commit I'd actually use: https://github.com/quietvoid/xbmc/commit/4561c3797ed95fbd9df5b311d6eb8ec5860cf4e9
I don't know if making another PR to merge into this one is acceptable here.

fuzzard commented:
Add it to this PR as just another commit.

quietvoid commented:
Seems `openssl-sys` needs to be pointed to the native depend, somehow.
Using `OPENSSL_DIR`.

quietvoid commented:
With the latest commit series, I've been able to build all supported Android platforms.

quietvoid commented:
`CBitstreamConverter` indeed doesn't seem to be the right place for the Dolby Vision NAL manipulation, as it's mostly used for extradata.
In my case I don't care if `CBitstreamConverter::Open` fails, so I probably want to do a separate processor.

sscobici commented:
@quietvoid thank you for this PR. I was thinking about Kodi's ability to play DV .m2ts which has 2 hevc streams, but it would be another quite a big PR.
Do you think ffmpeg would also like this option while playing DV FEL files? Maybe this code change should be there as a filter/demuxer option?

quietvoid commented:
So a quick question on this PR, would there be a chance to merge it in its current state?
The basic is that it works but some files don't go through `CBitstreamConverter` so the conversion isn't guaranteed.

I think CoreELEC has made a couple changes on their side but otherwise it works for them as well.

da-anda commented:
@quietvoid first off, I'm not a dev, so can't judge if this PR can/should be merged in it's current state, sorry. I was wondering though if a setting is really needed for this. I mean, why wouldn't a user want to convert an incompatible DoVi format to something his device does support? Or do we need this setting because we can not detect which DoVi layers are supported by the device? In the later case, do you think a multi-select setting where the user can select the supported DoVi profiles would make more sense and we only convert if current profile is not within the list?

quietvoid commented:
> I mean, why wouldn't a user want to convert an incompatible DoVi format to something his device does support?

Extra overhead, I guess.
Some users could want to leave it untouched even if it means inaccurate image output.

> Or do we need this setting because we can not detect which DoVi layers are supported by the device? In the later case, do you think a multi-select setting where the user can select the supported DoVi profiles would make more sense and we only convert if current profile is not within the list?

We can't reliably detect what profiles are supported.
I don't think a profile list would make sense since the conversion is already only active for one profile, which is profile 7.

thexai commented:
Runtime tested in Shield Pro 2019 (rebased locally on top of current master) and no issues found testing DV profile 7 videos.

Some considerations:

- Shield is not the best device to test this because already supports playback of DV profile 7 without convert. Likewise, the test is valid to look for possible regressions or side effects (nothing found until now).
- I had already tried it before, months ago, and the results were not good: random black screen after resume or seek on some files (not reproducible on all files). Even hard crash and entire Shield restart. I don't really know what it was, I think this was with FFmpeg 5.x so many things has changed. In any case, those problems seems are no longer present. Some users has reported similar issues in forum https://forum.kodi.tv/showthread.php?tid=371557&pid=3156335#pid3156335 but maybe are using old builds.

My personal opinion is that it is worth merge it so that it can be tested more and better. Let's not forget that v21 is still in alpha and therefore the best time to merge is probably now.

da-anda commented:
> We can't reliably detect what profiles are supported.
> I don't think a profile list would make sense since the conversion is already only active for one profile, which is profile 7.

@quietvoid can we detect if a DV capable display is attached or if DV in general is enabled on Android (like how we enumerate supported display resolutions, audio formats etc)? Just asking because our settings system has the possibility to display settings conditionally, and if we can enumerate general DV capabilities on Android, we could only show this setting if supported by display/hw to not confuse users.

PS: thanks for dovi_tool :)

thexai commented on tools/depends/native/cargo-c/CARGO-C-VERSION:
```diff
@@ -0,0 +1,5 @@
+APPNAME=cargo-c
+BASE_URL=https://github.com/lu-zero/cargo-c/archive
+VERSION=v0.9.20
+ARCHIVE=$(VERSION).tar.gz
```
Missing file name in `android-tools/xbmc-tarballs`

To follow the "pattern":

```
APPNAME=cargo-c
VERSION=v0.9.20
SOURCE=$(APPNAME)-$(VERSION)
ARCHIVE=$(SOURCE).tar.gz
```

Same in others

quietvoid replied:
Not sure where SOURCE is used but since we download off of tags, this won't work as the archive is only named by version.

thexai replied:
Maybe I have not explained myself correctly and also Linux depends build is not my thing but what I want to say is that all the other file names have the package name and the version and I don't think it has to be different for rust and cargo:

![tarballs](https://github.com/xbmc/xbmc/assets/58434170/44bf55df-6517-4ce0-adf7-d777fe04e145)

> Not sure where SOURCE is used

`SOURCE` is only a aux. variable to construct filename with `APPNAME`-`VERSION` format

Maybe @fuzzard can explain better

thexai replied:
As example all others follow same pattern:

https://github.com/xbmc/xbmc/blob/4dcaf8634c8519fb946557704831a991205955f1/tools/depends/native/cmake/Makefile#L5-L9

quietvoid replied:
I understand but you suggested changing `ARCHIVE`, which will break downloading off of GitHub tags.  

Other dependencies are uploaded to the Kodi source repo so they are not the same.
They don't use `BASE_URL`.

thexai replied:
I understand now. So probably until it is merged and the sources are uploaded to mirrors (and properly named), it cannot be done. Forget it for now.

thexai commented on tools/depends/target/libdovi/LIBDOVI-VERSION:
```diff
@@ -0,0 +1,6 @@
+LIBNAME=libdovi
+BASE_URL=https://github.com/quietvoid/dovi_tool/archive
+VERSION=libdovi-3.1.2
```
Here version number only and then:

`SOURCE=$(LIBNAME)-$(VERSION)`

da-anda commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -23453,3 +23453,15 @@ msgstr ""
+#. Title of Dolby Vision RPU conversion setting
+#: system/settings/settings.xml
+msgctxt "#39195"
+msgid "Convert Dolby Vision for compatibility"
```
In order to make this setting sound a little less technical, how about: `Dolby Vision compatibility mode`. And for the description:
`If enabled, Dolby Vision profile 7 will be converted to profile 8.1, which is more commonly supported by devices. Enable if your device does support Dolby Vision in general, but has issues with some videos.`

PS: sorry for nit picking here. Feel free to ignore until a final merge decision has been made 🤞. Thanks.

quietvoid replied:
Done. Thanks!

quietvoid committed:
tools/depends: Add Rust toolchain, cargo-c and libdovi dependencies

quietvoid commented:
I've made some changes to the labels. Also added a condition on the container hints so that the conversion is enabled only when `m_hints.dovi.dv_profile == 7`.

At the time the hints didn't have the profile yet. This allows for less overhead in other profiles playback, since they don't need to be parsed to know the profile.
Assumes the container has accurate metadata, but it usually does.

> @quietvoid can we detect if a DV capable display is attached or if DV in general is enabled on Android (like how we enumerate supported display resolutions, audio formats etc)? Just asking because our settings system has the possibility to display settings conditionally, and if we can enumerate general DV capabilities on Android, we could only show this setting if supported by display/hw to not confuse users.

The capability is already in Kodi but I haven't looked into using complex conditions.
Seems somewhat simple though.

> edit: sorry for going way ahead, but could libdovi in future also be used to convert DV to HDR10+ and vice versa?

HDR10+ can be converted but the complexity is in adding the Dolby Vision RPU into the bitstream.
HDR10+ are prefix SEIs while Dolby Vision must be appended to the AU.

It would have to be kept in memory until the end of the AU and it's not something I looked at for now.
The parsing capabilities of `CBitstreamConverter` are quite limited.

thexai commented:
> The capability is already in Kodi but I haven't looked into using complex conditions.
> Seems somewhat simple though.

It seems that all the pieces are there, something like this should be enough:

```diff
diff --git a/xbmc/settings/SettingConditions.cpp b/xbmc/settings/SettingConditions.cpp
index 1b81756764..8056205de2 100644
--- a/xbmc/settings/SettingConditions.cpp
+++ b/xbmc/settings/SettingConditions.cpp
@@ -118,6 +118,19 @@ bool SupportsVideoSuperResolution(const std::string& condition,
   return CServiceBroker::GetWinSystem()->SupportsVideoSuperResolution();
 }
 
+bool SupportsDolbyVision(const std::string& condition,
+                         const std::string& value,
+                         const SettingConstPtr& setting,
+                         void* data)
+{
+  CHDRCapabilities caps = CServiceBroker::GetWinSystem()->GetDisplayHDRCapabilities();
+
+  if (caps.SupportsDolbyVision())
+    return true;
+
+  return false;
+}
+
 bool SupportsScreenMove(const std::string& condition,
                         const std::string& value,
                         const SettingConstPtr& setting,
@@ -466,6 +479,7 @@ void CSettingConditions::Initialize()
   m_complexConditions.emplace("hassystemsdrpeakluminance", HasSystemSdrPeakLuminance);
   m_complexConditions.emplace("supportsscreenmove", SupportsScreenMove);
   m_complexConditions.emplace("supportsvideosuperresolution", SupportsVideoSuperResolution);
+  m_complexConditions.emplace("supportsdolbyvision", SupportsDolbyVision);
   m_complexConditions.emplace("ishdrdisplay", IsHDRDisplay);
   m_complexConditions.emplace("ismasteruser", IsMasterUser);
   m_complexConditions.emplace("hassubtitlesfontextensions", HasSubtitlesFontExtensions);
```

quietvoid commented:
I wonder if I should remove `<requirement>HAS_MEDIACODEC</requirement>` now that it's limited by
`SupportsDolbyVision`.

quietvoid committed:
[cmake][android] Add libdovi as optional build dependency

quietvoid committed:
[Android] Add Dolby Vision compatibility mode setting

For Dolby Vision profile 7 video files, the metadata is converted automatically
to improve compatibility on certain Android devices

thexai commented:
The two conditions are fine because this is a Android only feature and `<requirement>HAS_MEDIACODEC</requirement>` is the constant condition that assures this.

All seems ready now, I'm going to test the last update on Shield 🙂

thexai reviewed (approved):
Runtime tested on Shield 2019 and no issues found. All is smooth. 👍

New GUI setting is not visible when system not supports Dolby Vision, i.e from Shield settings disabling DV capability.

thexai commented:
@fuzzard 
Is it necessary to upload depends sources (rustup, cargo-c) to mirrors before merge or can it be done later?

fuzzard commented:
Yes put them on the mirrors.

With that said, if this is going to go in, is there really a need for an enable option? What's the purpose of not having it, build times only?
If yes, what is the actual increase in build times?

If it's negligible, remove optional and just force it

thexai commented:
The build time increase only is when depends are build (not Kodi) and in Android this only happens when some depends changes, then I think is not much relevant.

fuzzard commented:
It's absolutely relevant, as it then guides us on whether we need an enable option or not. Again, what's the point of the enable. Why would it ever be off? If the build time increase by a minute, remove it, if the build time increases by an hour, then it is an option to keep

quietvoid commented:
> Again, what's the point of the enable. Why would it ever be off?

Originally it was intended to be optional at compile time, defaulting to off.
Then for CI it had to be enabled by default, but that was meant to be temporary.

The library isn't useful for most non-ARM Android targets, except those supported by CoreELEC & co. that have Dolby Vision support through vendor kernels.

Otherwise  it depends if we want the Rust toolchain to always be setup even though there isn't anything else using it in the codebase at the moment.

On my system, the toolchain and libdovi compile takes about 1min30 total at most (~24 threads AMD CPU).
On the CI the logs say it's around a minute or so.

I should probably fix the rustup install to only do the minimal setup for `rustc`, could save some seconds.

fuzzard commented:
So it doesn't function on stock android?

quietvoid commented:
> So it doesn't function on stock android?

What do you mean? It works if the device supports Dolby Vision through MediaCodec.
Otherwise it simply wouldn't be used but it should still build/link and Kodi would run fine.

fuzzard commented:
> The library isn't useful for most non-ARM Android targets, except those supported by CoreELEC & co. that have Dolby Vision support through vendor kernels.

Trying to understand this. Non-arm, as in x86/x86_64 android?

quietvoid commented:
> Trying to understand this. Non-arm, as in x86/x86_64 android?

Yes, almost all the consumer Android devices supporting Dolby Vision are ARM.
I'm not sure if there are x86 ones at all.

I think I got confused with thinking CoreELEC did Dolby Vision on x86, but it actually seems to be ARM SoCs.

fuzzard commented:
Ok, so thats a non issue for us then. We dont actively distribute x86/x86_64, its only done as a CI target, and outside of @joseluismarti i dont think anyone even knows if kodi runs on those arches currently.

What im trying to narrow this down to, is optional stuff is bad in general. It then doesnt run CI, it doesnt get distributed in kodi releaases (nightly/release apk's, play store), so then the code "rots".  If the code works on regular android platforms (ie not custom kernels, not other custom shit, not only a single device in existence, etc), and someone wants to merge it, then build it as standard, and ship it as standard. If the feature doesnt add a crazy build time, ship it stock, remove the optional/enables. Enables are never tested outside of stock, so adding feature enables will just lead to maintenance/support hassles.

If it requires custom stuff (eg coreelec), then why merge into master, and they will just have to manage it downstream, otherwise the maintenance burden then gets shifted back to us. From what i understand, this isnt the case.

Just to say, im absolutely not against this, it doesnt affect me either way, but maintainability becomes an issue for the team at the best of times. Go all in, or not at all imo. if it works on a broad set of standard devices that users have that dont require tinkering, ship it.

quietvoid commented:
Given that x86 isn't really maintained, I agree that the optional flag should probably be removed.
I assume Kodi side it would probably be better in `PLATFORM_REQUIRED_DEPS`? Or should it stay optional there?

One more question before I make more changes:

Should the installed `rustc` version be made fixed?
Currently the latest stable is installed, so it may be different independently from Kodi's versioning.

fuzzard commented:
I think you can leave it as platform_optional, and then you dont have to change any of the define code in core, as its not absolutely required for kodi to function, and it reduces the effort for you (ie just remove the configure enable and stuff  around options in tools/depends).

As for fixed, i would lean towards yes. Again from a support side of things, rustc updates, and potentially breaks something because latest was downloaded by a user at some point in time just causes us grief. Just pick a version for now and if/when people want/need updated rustc for whatever, versions can be bumped later and breakages dealt when someone is actively maintaining that change.

quietvoid commented:
Removed the optional configure option. Now the Rust toolchain and `cargo-c` are always built for Android targets.

Updated the rustup install to use the minimal profile and to install fixed version (1.71.0 currently is the latest).
Also updated `cargo-c` but I don't know if the mirrors were already set.

fuzzard commented on tools/depends/target/libdovi/LIBDOVI-VERSION:
```diff
@@ -0,0 +1,7 @@
+LIBNAME=libdovi
+BASE_URL=https://github.com/quietvoid/dovi_tool/archive
+VERSION=3.1.2
```
@quietvoid im just putting sources on our mirrors. Ive added rustup and cargo-c, so if you can remove the BASE_URL from their version files.

However i just wanted to confirm you want this version still. I saw you released a 3.2.0 version last month. Let me know what version you want to roll with and i'll get it on mirrors so we can merge this.

quietvoid replied:
The only difference would be that some warnings are fixed, but no meaningful changes.

We could just bump everything as well, `cargo-c` and `libdovi` at least had new tags.
The Rust latest stable also got a new version, but it doesn't rely on mirrors.

fuzzard replied:
Ok, we can just go with 3.1.2, and always update later if you desire.

Can you remove the BASE_URL from this, CARGO-C-VERSION and RUSTUP-VERSION respectively and then we'll be good to merge

quietvoid replied:
Does `ARCHIVE` need to be renamed too? On the mirrors I presume they have the name now?
There were concerns before because the name was just the version (`BASE_URL` limitation), and the suggestion was to always do
```
SOURCE=$(APPNAME)-$(VERSION)
ARCHIVE=$(SOURCE).tar.gz
```

quietvoid commented:
Updated the depends files to use the Kodi mirrors.
It should work once `libdovi-3.1.2.tar.gz` is uploaded there.

abc123bac1 commented:
I had a quick question about the dolby vision compatibility mode introduced by this PR.

My understanding is that the EL is discarded to create a profile 8 dolby vision stream, but shouldn't this include the RPU information? I tried playing the BL_EL.mkv file from [here](https://mega.nz/folder/1eYDlZpS#tkj-)A74xfNixnFFZPcNqw) and do not see any RPU response (flashing colours from 26 seconds onwards0

If the RPU response is ignored, is the stream being played the same as an HDR10 stream, but it just triggers DV mode in the TV?

quietvoid commented:
> I had a quick question about the dolby vision compatibility mode introduced by this PR.
> 
> My understanding is that the EL is discarded to create a profile 8 dolby vision stream, but shouldn't this include the RPU information? I tried playing the BL_EL.mkv file from [here](https://mega.nz/folder/1eYDlZpS#tkj-GA74xfNixnFFZPcNqw) and do not see any RPU response (flashing colours from 26 seconds onwards0
> 
> If the RPU response is ignored, is the stream being played the same as an HDR10 stream, but it just triggers DV mode in the TV?

No, those files have some extra metadata that is removed by the profile 8.1 conversion.
That metadata causes the flashes but it's not valid without the EL.

abc123bac1 commented:
So is my understanding correct that what is being played is just a HDR10 stream with the TV being flipped into DV mode?

quietvoid commented:
> So is my understanding correct that what is being played is just a HDR10 stream with the TV being flipped into DV mode?

No, it's Dolby Vision profile 8.1.
